[wallet] [rpc] introduce 'label' API for wallet

jnewbery opened the pull request:
Add label API to wallet RPC.

This is one step towards #3816 ("Remove bolt-on account system") although it doesn't
actually remove anything yet.

These initially mirror the account functions, with the following differences:

- These functions aren't DEPRECATED in the help
- Help mentions 'label' instead of accounts. In the language used, labels are
  associated with addresses, instead of addresses associated with labels. (unlike
  with accounts.)
- Labels have no balance
  - No balances in `listlabels`
  - `listlabels` has no minconf or watchonly argument
- Like in the GUI, labels can be set on any address, not just receiving addreses
- Unlike accounts, labels can be deleted.
  Being unable to delete them is a common annoyance (see #1231).
  Currently only by reassigning all addresses using `setlabel`, but an explicit
  call `deletelabel` which assigns all address to the default label may make
  sense.

laanwj committed:
[wallet] [rpc] introduce 'label' API for wallet

Add label API to wallet RPC.

This is one step towards #3816 ("Remove bolt-on account system") although it doesn't
actually remove anything yet.

These initially mirror the account functions, with the following differences:

- These functions aren't DEPRECATED in the help
- Help mentions 'label' instead of accounts. In the language used, labels are
  associated with addresses, instead of addresses associated with labels. (unlike
  with accounts.)
- Labels have no balance
  - No balances in `listlabels`
  - `listlabels` has no minconf or watchonly argument
- Like in the GUI, labels can be set on any address, not just receiving addreses
- Unlike accounts, labels can be deleted.
  Being unable to delete them is a common annoyance (see #1231).
  Currently only by reassigning all addresses using `setlabel`, but an explicit
  call `deletelabel` which assigns all address to the default label may make
  sense.

Thanks to Pierre Rochard for test fixes.

jnewbery commented:
This is #7729 with the following changes:

- rebased on master
- fixed https://github.com/bitcoin/bitcoin/pull/7729#discussion_r177767690
- used `address` instead of `bitcoinaddress` as the argument for several RPC methods (https://github.com/bitcoin/bitcoin/pull/7729#discussion_r177768880)
- remove destdata from RPC return objects (https://github.com/bitcoin/bitcoin/pull/7729#discussion_r177769646)
- remove the `getlabel` RPC method, now that we have `getaddressinfo` (https://github.com/bitcoin/bitcoin/pull/7729#discussion_r177770329)
- sort labels before returning in `listlabels` RPC (https://github.com/bitcoin/bitcoin/pull/7729#discussion_r177771425)
- add an optional `force` bool argument to `getlabeladdress`, defaulting to False. When set to false, `getlabeladdress` will fail with an error if called on a label that doesn't exist. When set to true `getlabeladdress` will create a new label and new label address if called on a label that doesn't exist. I originally wanted to always throw an error if the label didn't exist (https://github.com/bitcoin/bitcoin/pull/7729/files/fef41789e8d7293ccfd859bc90fa0a3b2484752d#diff-df7d84ff2f53fcb2a0dc15a3a51e55ce), but apparently some minors depend on being able to create labels/accounts this way (https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-199444036)
- `getaddressesbylabel` throws an error if called for a label that doesn't exist (https://github.com/bitcoin/bitcoin/pull/7729#discussion_r177776131)
- don't mark `getaddressinfo` as deprecated (https://github.com/bitcoin/bitcoin/pull/7729#discussion_r177767690)

@laanwj doesn't have time to maintain #7729. I'd like to try to get this merged before v0.17.

jnewbery commented:
For full history of this PR, see #7729. Requesting review from reviewers of that PR: @Sjors , @MarcoFalke , @ryanofsky , @jonasschnelli , @instagibbs , @sipa , @luke-jr .

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -294,10 +314,10 @@ UniValue setlabel(const JSONRPCRequest& request)
-            + HelpExampleCli("setlabel", "\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\" \"tabby\"")
-            + HelpExampleRpc("setlabel", "\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\", \"tabby\"")
+            + HelpExampleCli("setlabel", "\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XZ\" \"tabby\"")
+            + HelpExampleRpc("setlabel", "\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XZ\", \"tabby\"")
```
Unrelated change.

jnewbery replied:
reverted

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -205,14 +210,15 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
         throw std::runtime_error(
             "getlabeladdress \"label\"\n"
-            "\nReturns the current Bitcoin address for receiving payments to this label.\n"
+            "\nReturns the current 'label address' for this label.\n"
```
Since this call gets even weirder, shouldn't we deprecate it in this PR?

jnewbery replied:
Aparently some miners require this functionality (https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-199444036).

I'd personally prefer to deprecate/remove this RPC method, but in the interest of making progress on this, let's maintain the current functionality and reconsider deprecation later.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -187,6 +187,11 @@ UniValue getnewaddress(const JSONRPCRequest& request)
     return EncodeDestination(dest);
 }
 
+void DeleteLabel(CWallet& wallet, std::string label)
```
Nit, move to `CWallet::DeleteLabel(const std::string& label)`? Or make it `static` for now?

PierreRochard replied:
Agree, this would be the only direct `CWalletDB` call from rpcwallet, seems as though it should go through `CWallet`

jnewbery replied:
Moved into `CWallet`

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -205,14 +210,15 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
-    if (request.fHelp || request.params.size() != 1)
+    if (request.fHelp || request.params.size() > 2)
         throw std::runtime_error(
             "getlabeladdress \"label\"\n"
```
Missing `( force )`.

jnewbery replied:
fixed

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -205,14 +210,15 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
     }
 
-    if (request.fHelp || request.params.size() != 1)
+    if (request.fHelp || request.params.size() > 2)
```
Missing `< 1`.

jnewbery replied:
fixed

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -224,6 +230,20 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
     // Parse the label first so we don't generate a key if there's an error
     std::string label = LabelFromValue(request.params[0]);
+    bool force = false;
+    if (!request.params[1].isNull())
```
Nit, add `{}`.

jnewbery replied:
added

jnewbery commented:
Thanks for the review @promag . Will address comments tomorrow.

PierreRochard commented on src/wallet/rpcwallet.cpp:
```diff
@@ -308,22 +328,24 @@ UniValue setlabel(const JSONRPCRequest& request)
 
     std::string label;
-    if (!request.params[1].isNull())
+    if (!request.params[1].isNull()) {
```
It seems odd to me that `label` is marked as a required argument for `setlabel` but if it's not provided then we default to an empty string instead of throwing an error. If we want to keep the behavior then `label` should be marked as optional.

jnewbery replied:
I've updated the behaviour to match the help text: the rpc call will fail if no label is provided.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            + HelpExampleRpc("getaddressesbylabel", "\"tabby\"")
+        );
+
+    LOCK2(cs_main, pwallet->cs_wallet);
```
done

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            + HelpExampleRpc("listlabels", "receive")
+        );
+
+    LOCK2(cs_main, pwallet->cs_wallet);
```
done

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    std::vector<std::pair<std::string, UniValue>> addresses;
+    for (const std::pair<CTxDestination, CAddressBookData>& item : pwallet->mapAddressBook) {
+        if (item.second.name == label) {
+            addresses.push_back(Pair(EncodeDestination(item.first), AddressBookDataToJSON(item.second, false)));
```
yes, done.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    LOCK2(cs_main, pwallet->cs_wallet);
+
+    std::string purpose;
+    if (request.params.size() > 0) {
```
done

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+
+    std::set<std::string> label_set;
+    for (const std::pair<CTxDestination, CAddressBookData>& entry : pwallet->mapAddressBook) {
+        if (purpose.empty() || entry.second.purpose == purpose){
```
fixed

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+
+    // Move label names to a list and sort before returning
+    std::list<std::string> label_list(label_set.begin(), label_set.end());
+    label_list.sort();
```
Yes, you're right. I've removed `label_list`

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    std::list<std::string> label_list(label_set.begin(), label_set.end());
+    label_list.sort();
+    UniValue ret(UniValue::VARR);
+    for (const std::string &name : label_list) {
```
fixed

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3849,118 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+
+    // Find all addresses that have the given label
+    std::vector<std::pair<std::string, UniValue>> addresses;
+    for (const std::pair<CTxDestination, CAddressBookData>& item : pwallet->mapAddressBook) {
```
I'll leave this functionality here for now. Can be refactored later if required.

jnewbery commented:
Addressed review comments from @promag and @PierreRochard . Are you able to re-review?

PierreRochard commented:
Yes, re-reviewing now

PierreRochard commented:
A behavior difference between the old `getaccountaddress` and the new aliased `getlabeladdress` is causing the `wallet_basic.py` to fail. I was able to fix it by adding `force` to the test, but I think it's the endpoint that should be changed if the goal is to not break the API with the aliasing.

`rpc_listtransactions.py` and `wallet_listreceivedby.py` are also failing, it may be due to the same root cause but I won't be able to confirm until tomorrow.

jnewbery commented:
Thanks for the fixes @PierreRochard! I've updated the tests as you suggested and force pushed.

PierreRochard commented:
Tested ACK 1f7f1dfd1c16cf0c9a700e2b262ffa8f6c6559cd

laanwj commented:
I'm still not convinced we really need `getlabeladdress` in the first place, but anyhow.
utACK https://github.com/bitcoin/bitcoin/pull/12892/commits/1f7f1dfd1c16cf0c9a700e2b262ffa8f6c6559cd

MarcoFalke commented:
Needs rebase

PierreRochard commented:
@jnewbery I agree that `getlabeladdress` should be removed, it has too much overlap with `getnewaddress` and miner-specific use cases are likely best implemented in the mining software (this suggestion seems reasonable: https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-206938418)

ryanofsky commented:
> Other reviewers of this and the previous PR (@promag, @PierreRochard, @Sjors, @MarcoFalke , @ryanofsky , @instagibbs , @sipa , @luke-jr) - please can you provide concept input into whether we should just remove getlabeladdress?

I think it would be better to keep this PR as focused as possible. The goal should just be to fill in the last missing bits of the label implementation, so it is easy to switch from accounts to labels and we can get rid of broken "account balance" code and code that tries to track what account a transaction is sent from. I don't think it makes sense to introduce additional differences between accounts and labels that will make migration more difficult than necessary.

The main problem I see with `getlabeladdress` is just the current documentation:

> Returns the current 'label address' for this label.

Instead of something explanatory like:

> Returns the default receiving address for this label. This will reset to a fresh address once there's a transaction that spends to it.

Maybe there are other problems with `getlabeladdress`, but I can't see where they have been mentioned.

Also, I think this PR needs release notes. Good documentation is probably the most important part of this change.

jnewbery committed:
[docs] Add release notes for wallet 'label' API.

jnewbery commented:
> I don't think it makes sense to introduce additional differences between accounts and labels that will make migration more difficult than necessary.

ok, I can see the merit in that. Let's keep this PR focused and we can squabble over removing `getlabeladdress` later.

I've updated the help text for `getlabeladdress` and added release notes.

Re-review would be very much appreciated at this point :smiley:

laanwj commented:
My intent was not for the `getlabeladdress` discussion to block this PR.

> getlabeladdress doesn't make much sense to me, and blurs the distinction that "labels are associated with addresses, instead of addresses associated with labels".

Another problem with it is that it requires keeping this whole `CAccount` administration in place. Part of the reason for this account to label refactor is simplify the wallet mess and to get rid of that.

Will re-review.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3811,6 +3843,111 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    UniValue ret(UniValue::VOBJ);
+    for (const std::pair<CTxDestination, CAddressBookData>& item : pwallet->mapAddressBook) {
+        if (item.second.name == label) {
+            ret.push_back(Pair(EncodeDestination(item.first), AddressBookDataToJSON(item.second, false)));
```
`pushKV`.

jnewbery replied:
fixed

promag reviewed (commented):
Will test. Some nits thou.

promag commented on doc/release-notes-pr12892.md:
```diff
@@ -0,0 +1,28 @@
+  as keys instead of an array of strings.
+
+Conceptually, labels are associated with addresses, instead of addresses
+associated with labels. (unlike with accounts.)
```
Nit,
```
associated with labels (unlike with accounts).
```

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -3643,6 +3643,12 @@ std::set<CTxDestination> CWallet::GetLabelAddresses(const std::string& label) co
     return result;
 }
 
+void CWallet::DeleteLabel(std::string label)
```
Agree that's better. Fixed.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3720,6 +3734,17 @@ UniValue DescribeWalletAddress(CWallet* pwallet, const CTxDestination& dest)
+{
+    UniValue ret(UniValue::VOBJ);
+    if (verbose) {
+        ret.push_back(Pair("name", data.name));
```
fixed

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3720,6 +3734,17 @@ UniValue DescribeWalletAddress(CWallet* pwallet, const CTxDestination& dest)
+    if (verbose) {
+        ret.push_back(Pair("name", data.name));
+    }
+    ret.push_back(Pair("purpose", data.purpose));
```
fixed

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3811,6 +3843,111 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            "1. \"purpose\"  (string, optional) Address purpose to list labels for ('send','receive'). An empty string is the same as not providing this argument.\n"
+            "\nResult:\n"
+            "[                      (json array of string)\n"
+            "  \"label\",  (string) Label name\n"
```
Nit, could have better alignment:
```
Result:
[                      (json array of string)
  "label",  (string) Label name
  ...
] 
```

jnewbery replied:
fixed

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3811,6 +3843,111 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    std::set<std::string> label_set;
+    for (const std::pair<CTxDestination, CAddressBookData>& entry : pwallet->mapAddressBook) {
+        if (purpose.empty() || entry.second.purpose == purpose) {
+            label_set.insert(entry.second.name);
```
Build `ret` here:
```cpp
if (label_set.insert(entry.second.name).second) {
    ret.push_back(name);
}

jnewbery replied:
Placing into a set sorts the labels by name (`mapAddressBook` will be sorted by key, which is the `CTxDestination`)

jnewbery replied:
Added a comment to explain that this is to sort the labels.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -207,14 +206,15 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
             "\nArguments:\n"
-            "1. \"label\"         (string, required) The label name for the address. It can also be set to the empty string \"\" to represent the default label. The label does not need to exist, it will be created and a new address created  if there is no label by the given name.\n"
+            "1. \"label\"         (string, required) The label for the address. It can also be set to the empty string \"\" to represent the default label.\n"
+            "2. \"force\"         (bool, optional) Whether the label should be created if it does not yet exist. If False, the RPC will return an error if called with a label that doesn't exist. Defaults to False.\n"
```
`Defaults to` depends on the method.

ryanofsky replied:
> Defaults to depends on the method.

I guess simplest way to fix this would be to say "Defaults to false, unless the getaccountaddress method alias is being called, in which case the default is true for backwards compatibility."

If you wanted to be more clever I guess you could adapt the help text.

All of this is making me less a fan of the original suggestion to have a force argument, though:
https://github.com/bitcoin/bitcoin/pull/7729/files/fef41789e8d7293ccfd859bc90fa0a3b2484752d#r177781021

I definitely wouldn't object if you wanted to drop it.

jnewbery replied:
thanks! fixed

jnewbery replied:
> I guess simplest way to fix this would be to say...

Yes, that's clear. Updated.

> All of this is making me less a fan of the original suggestion to have a force argument, though.

Please - can we not hold this PR up on the details of the `getlabeladdress` RPC? I would much rather we merged something that made progress towards removing the accounts system, and then worked out what to do about `getlabeladdress` later.

promag reviewed (commented):
Looks good, more comments 😄

In some calls, if the response is empty, `RPC_WALLET_INVALID_LABEL_NAME ` is raised instead of just returning `[]`. However calling `listlabels` with an unknown purpose will return `[]`. Consistency would be nice (I prefer the empty array, I don't see it as an error).

The format of some help messages could be fixed, but I guess it's not relevant at this point.

BTW, I also don't see the value of `getlabeladdress`. Maybe just keep the old `getaccountaddress` under a deprecation flag?

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -309,23 +324,22 @@ UniValue setlabel(const JSONRPCRequest& request)
-    // Only add the label if the address is yours.
     if (IsMine(*pwallet, dest)) {
-        // Detect when changing the label of an address that is the 'unused current key' of another label:
+        // Detect when changing the label of an address that is the 'label address' of another label:
```
Maybe say receiving address instead of 'label address' here too.

jnewbery commented:
Thanks @promag and @ryanofsky - I've answered/addressed your comments.

> In some calls, if the response is empty, RPC_WALLET_INVALID_LABEL_NAME is raised instead of just returning []. However calling listlabels with an unknown purpose will return [].

I don't think that's inconsistent - calling eg `getaddressesbylabel` with a non-existent label throws an error. Calling `listlabels` with a purpose that no addresses are using returns the list of all addresses that have that purpose (ie `[]`).

> I also don't see the value of getlabeladdress

You and me both pal, but let's defer that discussion for a future PR.

ryanofsky commented on doc/release-notes-pr12892.md:
```diff
@@ -0,0 +1,28 @@
+'label' API for wallet
+----------------------
+
+A new 'label' API has been introduced for the wallet. This is intended as a
```
This is great! There are a lot of changes I would suggest but I thought it would be easier to just make them in a wiki page:

https://github.com/bitcoin-core/bitcoin-devwiki/wiki/12892-Label-API-release-notes

The main change is trying to organize the RPC information in a table.

Feel free to delete the wiki page and move anything here.

ryanofsky reviewed (commented):
utACK 883605dc4c0d28f32b724a2cf46fee87711ab6e9. Just for the record, the reason I hadn't posted an ack earlier was because I wanted to go through the release notes, not because of https://github.com/bitcoin/bitcoin/pull/12892#discussion_r180464405. But I get that #7729 was open for 2 years, so you might be trying to make up lost time. :stuck_out_tongue:

jnewbery commented:
Thanks for the release notes suggestions Russ. I've taken them almost exactly verbatim from your wiki, with the following change:

> delete the previous label associated with an address, instead of making an implicit getaccountaddress call to ensure the previous label still has a receiving address

to

> delete the previous label associated with an address when the final address using that label is reassigned to a different label, instead of making an implicit getaccountaddress call to ensure the previous label still has a receiving address

ryanofsky reviewed (commented):
utACK cc505e493ac06901776332a538cde5624e83fa84. I made a few very minor tweaks to punctuation and wording in the wiki, but feel free to ignore these unless perhaps you're updating the PR anyway.

jnewbery commented:
> feel free to ignore these unless perhaps you're updating the PR anyway.

Thanks, let's leave these for now. Minor punctuation/wording can be fixed when release notes are merged at the end of the release cycle.

jnewbery commented:
> feel free to ignore these unless perhaps you're updating the PR anyway.

Bleurgh. lint-whitespace didn't like my release notes. Fixed them and pulled in Russ's changes.

jamesob commented:
utACK https://github.com/bitcoin/bitcoin/pull/12892/commits/b00faff9ed653ea28d02ff6d821183389fcd8b29

Looks good. The release notes are very helpful.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3811,6 +3844,112 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+        purpose = request.params[0].get_str();
+    }
+
+    // Add to a list to sort by label name, then insert into Univalue array
```
`Add to a set ...`?

Anyway, I mean there is no point in sorting the response, could very well be an unsorted list. We just want to make sure there are no repeated labels.

jnewbery replied:
Thanks. Fixed to `Add to a set`

promag reviewed (commented):
utACK b00faff.

jnewbery commented:
Fixed comment. b00faff -> 41ba061

kallewoof reviewed (commented):
utACK 41ba0618048b7b2d7e257bb52cf8f138d29dabd3

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -226,6 +227,21 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
 
     // Parse the label first so we don't generate a key if there's an error
     std::string label = LabelFromValue(request.params[0]);
+    bool force = request.strMethod == "getaccountaddress" ? true : false;
```
duh. Will tidy up in follow-up PR.

promag replied:
Let it go, will go away when accounts are removed.

kallewoof replied:
Yeah, not worth a PR for just that.

jnewbery replied:
Fixed in #12953
